src/test/scala/chiselTests/CompatibilityInteroperabilitySpec.scala

```diff
@@ -390,4 +390,137 @@ class CompatibilityInteroperabilitySpec extends ChiselFlatSpec {
+        val bar = Output(Bool())
+      }
+
+    }
+    object Chisel3 {
+      import chisel3._
+      import chisel3.util.Valid
+      import chisel3.experimental.hierarchy.{instantiable, public, Instance}
+
+      class FooModuleIO extends Bundle {
+
+        val quz = Input(new QuzIO)
+        val foo = Output(Bool())
+        val bar = Input(Bool())
+      }
+
+      class QuzIO extends Bundle {
+        val q = Flipped(Valid(new Compat.LegacyChiselIO))
+      }
+
+      @instantiable
+      class FooModule extends Module {
+
+        @public
+        val io = IO(new FooModuleIO())
+        io <> DontCare
+      }
+
+      class FooMirrorBlackBox extends BlackBox {
+        val io = IO(Flipped(new FooModuleIO()))
+      }
+
+      class Top() extends Module {
+
+        val foo = Module(new FooModule)
+        val mirror = Module(new FooMirrorBlackBox())
+        foo.io <> mirror.io
+        foo.io <> DontCare
+      }
+    }
+
+    compile(new Chisel3.Top)
+  }
+
+  "A chisel3.Bundle connected to a Chisel bundle in either direction" should "error with mismatched fields" in {
+    object Compat {
+      import Chisel._
+      // Chisel._ bundle
+      class FooBundle extends Bundle {
+        val foo = UInt(width = 8)
+      }
+    }
+    object Chisel3 {
+      import chisel3._
+      class BarBundle extends Bundle {
+        val bar = UInt(8.W)
+      }
+
+      class MyModule(swap: Boolean) extends Module {
+        val in = IO(Input(if (swap) new Compat.FooBundle else new BarBundle))
+        val out = IO(Output(if (swap) new BarBundle else new Compat.FooBundle))
+        out <> in
+      }
+    }
+    val thrown0 = the[chisel3.internal.ChiselException] thrownBy (compile(new Chisel3.MyModule(true)))
+    thrown0.getMessage should include("Left Record missing field (foo)")
+    val thrown1 = the[chisel3.internal.ChiselException] thrownBy (compile(new Chisel3.MyModule(false)))
+    thrown1.getMessage should include("Left Record missing field (bar)")
+  }
+  it should "error with missing fields in the Chisel._" in {
+    object Compat {
+      import Chisel._
+      // Chisel._ bundle
+      class FooBundle extends Bundle {
+        val foo = UInt(width = 8)
+      }
+    }
+    object Chisel3 {
+      import chisel3._
+      class FooBarBundle extends Bundle {
+        val foo = UInt(8.W)
+        val bar = UInt(8.W)
+      }
+
+      class MyModule(swap: Boolean) extends Module {
+        val in = IO(Input(if (swap) new Compat.FooBundle else new FooBarBundle))
+        val out = IO(Output(if (swap) new FooBarBundle else new Compat.FooBundle))
+        out <> in
+      }
+    }
+    val thrown0 = the[chisel3.internal.ChiselException] thrownBy (compile(new Chisel3.MyModule(true)))
+    thrown0.getMessage should include("Right Record missing field (bar)")
+    val thrown1 = the[chisel3.internal.ChiselException] thrownBy (compile(new Chisel3.MyModule(false)))
+    thrown1.getMessage should include("Left Record missing field (bar)")
+  }
```

Comment:
I'm not against this stricter behavior in principle (we are try to get rid of compatibility mode after all), but this change makes this PR much more than a bug fix. This exact test (at least with `false` as the argument) works on Chisel 3.5.4: https://scastie.scala-lang.org/tdeqjGZqTMWXt3Swi4Nozg

Reply:
i can easily make it work if that is the behavior we want. I wasn't sure what the behavior was supposed to be

Reply:
it also works with true (as long as you initialize) https://scastie.scala-lang.org/jQaddfnrQaaE5Ar4TP9pzQ

Reply:
so I will relax the check so that as long as one of the sides of the connection is Chisel you are allowed to have mismatched fields

Reply:
Yeah that sounds reasonable. We'll be tightening up the behavior here when we delete compatibility mode so keeping this zombie alive a little bit longer is the right move for now (esp. for a bug fix).

Reply:
relaxed and updated the test